django/core/management/commands/shell.py

```diff
@@ -42,25 +42,8 @@ def bpython(self, options):
     def python(self, options):
         import code
 
-        # Set up a dictionary to serve as the environment for the shell, so
-        # that tab completion works on objects that are imported at runtime.
+        # Set up a dictionary to serve as the environment for the shell.
         imported_objects = {}
-        try:  # Try activating rlcompleter, because it's handy.
-            import readline
-        except ImportError:
-            pass
-        else:
-            # We don't have to wrap the following import in a 'try', because
-            # we already know 'readline' was imported successfully.
-            import rlcompleter
-            readline.set_completer(rlcompleter.Completer(imported_objects).complete)
-            # Enable tab completion on systems using libedit (e.g. macOS).
-            # These lines are copied from Python's Lib/site.py.
-            readline_doc = getattr(readline, '__doc__', '')
-            if readline_doc is not None and 'libedit' in readline_doc:
-                readline.parse_and_bind("bind ^I rl_complete")
```

Comment:
You appear to have dropped support for macOS.

Reply:
How so?

Reply:
Unless I misread (maybe I misread!), you removed the four lines of code above this comment that refers to libedit without adding it back somewhere else.

Reply:
This block of code came from default value of `sys.__interactivehook__` (https://github.com/python/cpython/blob/master/Lib/site.py#L441-L487).  Since we now call `sys.__interactivehook__()`, that block can be removed.

But I've only been able to test on linux, so it would be very helpful if you could test on macOS.

Reply:
I see. Anyway this is still a breaking change. The function to which you link, `site.enablerlcompleter`, is called [only if](https://docs.python.org/3/library/site.html#readline-configuration) Python was started [without](https://github.com/python/cpython/blob/eeb701adc0fc29ba803fddf133d917ff45639a00/Lib/site.py#L614-L615) [`-S`] and [without](https://github.com/python/cpython/blob/eeb701adc0fc29ba803fddf133d917ff45639a00/Lib/site.py#L606-L607) [`-I`] (see also [`sys.flags`]), so it doesn't always set `sys.__interactivehook__` by default. The prior behavior of the `shell` command` enabled the rlcompleter with macOS support unconditionally. Your patch now only enables macOS support under some operating conditions.

Perhaps it would be better to lean directly on the `site` module. I'm not sure it's a good idea to call `site.enablerlcompleter`: On the one hand, it's not documented. On the other, the name doesn't start with an underscore. One option might be unconditionally calling [`site.main`](https://docs.python.org/3/library/site.html#site.main) so that user and site customization script behavior matches vanilla Python.

[`-S`]: https://docs.python.org/3/using/cmdline.html#id3
[`-I`]: https://docs.python.org/3/using/cmdline.html#id2
[`sys.flags`]: https://docs.python.org/3/library/sys.html#sys.flags

Reply:
Thanks for your input.

> Anyway this is still a breaking change... Your patch now only enables macOS support under some operating conditions.

I'm not sure about that, because as far as I can see these operating conditions are only ever theoretical.

If you start Python with `-S`, then you can't load the shell at all:

```
$ python -S manage.py shell
Traceback (most recent call last):
  File "manage.py", line 11, in main
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 22, in <module
    main()
  File "manage.py", line 13, in main
    raise ImportError(
ImportError: Couldn't import Django. Are you sure it's installed and available on your PYTHONPATH environment variable? Did you forget to activate a virtual environment?
```

And if you start it with `-i`, you can start the shell, but you can't import any of your apps:

```
$ python -I manage.py shell
Python 3.8.2 (default, May  2 2020, 15:51:29) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import myapp
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ModuleNotFoundError: No module named 'myapp'
```

I suppose you could argue there is a breaking change if you're somehow relying on an overridden `sys.__interactivehook__()` _not_ being called, but I'm struggling to come up with why you might do that.

> I'm not sure it's a good idea to call `site.enablerlcompleter`: On the one hand, it's not documented. On the other, the name doesn't start with an underscore.

We're not calling `site.enablerlcompleter`, but instead we're calling whatever the value of `sys.__interactivehook__` is, and [that is documented](https://docs.python.org/3/library/sys.html#sys.__interactivehook__).  (I know it begins with an underscore, but I think that's OK.)

> One option might be unconditionally calling `site.main` so that user and site customization script behavior matches vanilla Python.

That's a nice idea, but by the time we get to the code that starts the Django shell, it will already have been called.  In any case, it calls `enablerlcompleter`, which _sets_ `sys.__interactivehook__`.  Something still needs to _call_ `sys.__interactivehook__`, and in normal Python shell, this is done by [code in `main.c`](https://github.com/python/cpython/blob/c9b8e9c421b57acdcaf24fab0c93bc29b3ef7c67/Modules/main.c#L441-L472), which as far as I can tell we can never access from Python.

Reply:
> these operating conditions are only ever theoretical.

They're not. I use them all the time together with [`shell`]'s [`--pythonpath`] option. (And I do this on a Mac, hence my interest in making sure we get this PR right.)

[`--pythonpath`]: https://docs.djangoproject.com/en/3.1/ref/django-admin/#cmdoption-pythonpath
[`shell`]: https://docs.djangoproject.com/en/3.1/ref/django-admin/#shell